cmd/snap-seccomp/main.go

```diff
@@ -0,0 +1,667 @@
+	sc.arch = C.htot32(C.__u32(arch), C.__u32(arch))
+}
+func (sc *SeccompData) SetArgs(args [6]uint64) {
+	for i := range args {
+		sc.args[i] = C.htot64(sc.arch, C.__u64(args[i]))
+	}
+}
+
+func readNumber(token string) (uint64, error) {
+	if value, ok := seccompResolver[token]; ok {
+		return value, nil
+	}
+
+	// Negative numbers are not supported yet, but when they are,
+	// adjust this accordingly
+	return strconv.ParseUint(token, 10, 64)
+}
+
+func parseLine(line string, secFilter *seccomp.ScmpFilter) error {
+	line = strings.TrimSpace(line)
+
+	// ignore comments and empty lines
+	if strings.HasPrefix(line, "#") || line == "" {
+		return nil
+	}
+
+	// regular line
+	tokens := strings.Fields(line)
+	if len(tokens[1:]) > ScArgsMaxlength {
+		return fmt.Errorf("too many arguments specified for syscall '%s' in line %q", tokens[0], line)
+	}
+
+	// fish out syscall
+	secSyscall, err := seccomp.GetSyscallFromName(tokens[0])
+	if err != nil {
+		// FIXME: use structed error in libseccomp-golang when
+		//   https://github.com/seccomp/libseccomp-golang/pull/26
+		// gets merged. For now, ignore
+		// unknown syscalls
+		return nil
+	}
+
+	var conds []seccomp.ScmpCondition
+	for pos, arg := range tokens[1:] {
+		var cmpOp seccomp.ScmpCompareOp
+		var value uint64
+		var err error
+
+		if arg == "-" { // skip arg
+			continue
+		}
+
+		if strings.HasPrefix(arg, ">=") {
+			cmpOp = seccomp.CompareGreaterEqual
+			value, err = readNumber(arg[2:])
+		} else if strings.HasPrefix(arg, "<=") {
+			cmpOp = seccomp.CompareLessOrEqual
+			value, err = readNumber(arg[2:])
+		} else if strings.HasPrefix(arg, "!") {
+			cmpOp = seccomp.CompareNotEqual
+			value, err = readNumber(arg[1:])
+		} else if strings.HasPrefix(arg, "<") {
+			cmpOp = seccomp.CompareLess
+			value, err = readNumber(arg[1:])
+		} else if strings.HasPrefix(arg, ">") {
+			cmpOp = seccomp.CompareGreater
+			value, err = readNumber(arg[1:])
+		} else if strings.HasPrefix(arg, "|") {
+			cmpOp = seccomp.CompareMaskedEqual
+			value, err = readNumber(arg[1:])
+		} else {
+			cmpOp = seccomp.CompareEqual
+			value, err = readNumber(arg)
+		}
+		if err != nil {
+			return fmt.Errorf("cannot parse token %q (line %q)", arg, line)
+		}
+
+		var scmpCond seccomp.ScmpCondition
+		if cmpOp == seccomp.CompareMaskedEqual {
+			scmpCond, err = seccomp.MakeCondition(uint(pos), cmpOp, value, value)
+		} else {
+			scmpCond, err = seccomp.MakeCondition(uint(pos), cmpOp, value)
+		}
+		if err != nil {
+			return fmt.Errorf("cannot parse line %q: %s", line, err)
+		}
+		conds = append(conds, scmpCond)
+	}
+
+	// Default to adding a precise match if possible. Otherwise
+	// let seccomp figure out the architecture specifics.
+	if err = secFilter.AddRuleConditionalExact(secSyscall, seccomp.ActAllow, conds); err != nil {
+		err = secFilter.AddRuleConditional(secSyscall, seccomp.ActAllow, conds)
+	}
+
+	return err
+}
+
+// used to mock in tests
+var (
+	archUbuntuArchitecture       = arch.UbuntuArchitecture
```

Comment:
All of these "Ubuntu" names should at some point probably be replaced by something that reflects better the cross-distribution nature of this code.